sql/binlog.cc

```diff
@@ -5818,7 +5848,70 @@ int MYSQL_BIN_LOG::purge_index_entry(THD *thd, ulonglong *decrease_log_space,
     /* Get rid of the trailing '\n' */
     log_info.log_file_name[length - 1] = 0;
 
-    if (!mysql_file_stat(m_key_file_log, log_info.log_file_name, &s, MYF(0))) {
+    if ((error = find_log_pos(&check_log_info, log_info.log_file_name,
+                              need_lock_index))) {
+      if (error != LOG_INFO_EOF) {
+        if (thd) {
+          push_warning_printf(thd, Sql_condition::SL_WARNING,
+                              ER_BINLOG_PURGE_FATAL_ERR,
+                              "a problem with deleting %s and "
+                              "reading the binlog index file",
+                              log_info.log_file_name);
+        } else {
+          LogErr(INFORMATION_LEVEL,
+                 ER_BINLOG_CANT_DELETE_FILE_AND_READ_BINLOG_INDEX,
+                 log_info.log_file_name);
+        }
+        break;
+      }
+
+      error = 0;
+      if (!need_lock_index) {
+        /*
+          This is to avoid triggering an error in NDB.
+
+          @todo: This is weird, what does NDB errors have to do with
+          need_lock_index? Explain better or refactor /Sven
+        */
+        ha_binlog_index_purge_file(current_thd, log_info.log_file_name);
+      }
+
+      delete_list.push_back(std::string(log_info.log_file_name));
+    }
+  }
+
+  if (!error)
+    error = purge_logs_in_list(delete_list, thd, decrease_log_space,
+                               need_lock_index);
+
+err:
+  DBUG_RETURN(error);
+}
+
+/**
+  Deletes logs sepecified in a list if they exist on the file system
+  @param delete_list         The list of log files to delete
+  @param thd                 Pointer to the THD object
+  @param decrease_log_space  Amount of space freed
+  @param need_lock_index     Need to lock the index?
+  @retval
+    0                        ok
+  @retval
+    LOG_INFO_EMFILE          Too many files opened
+    LOG_INFO_FATAL           If any other than ENOENT error from
+                             mysql_file_stat() or mysql_file_delete()
+*/
+int MYSQL_BIN_LOG::purge_logs_in_list(std::list<std::string> &delete_list,
```

Comment:
`int MYSQL_BIN_LOG::purge_logs_in_list(const log_file_name_container& delete_list, ...`

See below for `log_file_name_container`